### PR TITLE
feat(wandb): offset logging step when resuming

### PR DESCRIPTION
**Pull request by borisdayma**
## What does this PR do?

When resuming a run, the step was not being offset to account for previous logged data.
It will now be correctly offset.

In addition, I took advantage to give users a hint when they have related issues trying to log "back in time". It is usually due to them logging manually other data (images, etc) and causing the wandb step to automatically increase. This will close https://github.com/wandb/client/issues/1507

I added the same details in the documentation.

## Before submitting
- [X] Was this discussed/approved via a Github issue? discussed in wandb issue
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [X] Did you make sure your PR does only one thing, instead of bundling different changes together? Since the other modification is just a warning
- [X] Did you make sure to update the documentation with your changes?
- [X] Did you write any new necessary tests? 
- [X] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loggers/wandb.py [hunk 1]
```diff
@@ -154,6 +161,9 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
         assert rank_zero_only.rank == 0, 'experiment tried to log from global_rank != 0'
 
         metrics = self._add_prefix(metrics)
+        if not WandbLogger.WARNING_DISPLAYED and (step is not None and step + self._step_offset < self.experiment.step):
+            log.warning('Trying to log at a previous step. Use `commit=False` if logging metrics manually.')
```
Comment (awaelchli) [inline pytorch_lightning/loggers/wandb.py:165]:
```suggestion
            log.warning('Trying to log at a previous step. Use `commit=False` when logging metrics manually.')
```
not sure :)

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loggers/wandb.py (hunk 1 above)
Comment (borisdayma) [reply pytorch_lightning/loggers/wandb.py]: Thanks!

File: pytorch_lightning/loggers/wandb.py [hunk 2]
```diff
@@ -154,6 +161,9 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
         assert rank_zero_only.rank == 0, 'experiment tried to log from global_rank != 0'
 
         metrics = self._add_prefix(metrics)
+        if not WandbLogger.WARNING_DISPLAYED and (step is not None and step + self._step_offset < self.experiment.step):
+            log.warning('Trying to log at a previous step. Use `commit=False` when logging metrics manually.')
+            WandbLogger.WARNING_DISPLAYED = True
```
Comment (Borda) [inline pytorch_lightning/loggers/wandb.py:166]:
```suggestion
        if not self.WARNING_DISPLAYED and (step is not None and step + self._step_offset < self.experiment.step):
            log.warning('Trying to log at a previous step. Use `commit=False` when logging metrics manually.')
            self.WARNING_DISPLAYED = True
```

**Review by Borda — APPROVED** [review]
can we add a test for the multiple call?

**Comment by borisdayma** [comment]
> can we add a test for the multiple call?

Sure, you're talking about the warning?

**Comment by borisdayma** [comment]
@Borda I added a test for warnings displayed.

File: pytorch_lightning/loggers/wandb.py [hunk 3]
```diff
@@ -154,6 +161,9 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
         assert rank_zero_only.rank == 0, 'experiment tried to log from global_rank != 0'
 
         metrics = self._add_prefix(metrics)
+        if not self.WARNING_DISPLAYED and (step is not None and step + self._step_offset < self.experiment.step):
```
Comment (tchaton) [inline pytorch_lightning/loggers/wandb.py:164]: I think we have a better way to handle only once displayed Warning. Could be WarningCache. Need to check.

Comment (Borda) [reply pytorch_lightning/loggers/wandb.py:164]: shall we reimplement this one?
https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/13863ccf1f96cb912406c029c50f4253f4599aee/pl_bolts/utils/warnings.py#L7

**Comment by borisdayma** [comment]
I'm now using `WarningCache` as suggested by @tchaton but it breaks my tests.
In particular I'm having trouble capturing the warnings. I tried with caplog, capsys and capfd but nothing seems to appear, even if I just do a direct call within the test to `warnings.warn` or even `logging.log`.

**Comment by borisdayma** [comment]
I was able to make the tests work with `WarningCache`.
I had to rely on `recwarn` instead of `caplog` (not sure why) and made a little helper function.

**Review by rohitgr7 — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Pull request merged**